src/net/ossl.cc

```diff
@@ -0,0 +1,2073 @@
+        if (!ssl_ctx) {
+            throw make_ossl_error(
+              "Failed to initialize SSL context");
+        }
+        const auto& ck_pair = _creds->get_certkey_pair();
+        if (type == session_type::SERVER) {
+            if (!ck_pair) {
+                throw make_ossl_error(
+                  "Cannot start session without cert/key pair for server");
+            }
+            switch (_creds->get_client_auth()) {
+            case client_auth::NONE:
+            default:
+                SSL_CTX_set_verify(ssl_ctx.get(), SSL_VERIFY_NONE, nullptr);
+                break;
+            case client_auth::REQUEST:
+                SSL_CTX_set_verify(ssl_ctx.get(), SSL_VERIFY_PEER, nullptr);
+                break;
+            case client_auth::REQUIRE:
+                SSL_CTX_set_verify(
+                  ssl_ctx.get(),
+                  SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT,
+                  nullptr);
+                break;
+            }
+
+            auto options = SSL_OP_ALL | SSL_OP_ALLOW_CLIENT_RENEGOTIATION;
+            if (_creds->is_server_precedence_enabled()) {
+                options |= SSL_OP_CIPHER_SERVER_PREFERENCE;
+            }
+
+            SSL_CTX_set_options(ssl_ctx.get(), options);
+
+            switch(_creds->get_session_resume_mode()) {
+                case session_resume_mode::NONE:
+                    SSL_CTX_set_session_cache_mode(ssl_ctx.get(), SSL_SESS_CACHE_OFF);
+                    break;
+                case session_resume_mode::TLS13_SESSION_TICKET:
+                    // By default, SSL contexts have server size cache enabled
+                    if (1 != SSL_CTX_set_tlsext_ticket_key_evp_cb(ssl_ctx.get(), &session_ticket_cb)) {
+                        throw make_ossl_error("Failed to set session ticket callback function");
+                    }
+                    break;
+            }
+        } else {
+            if (_creds->is_server_precedence_enabled()) {
+                SSL_CTX_set_options(ssl_ctx.get(), SSL_OP_CIPHER_SERVER_PREFERENCE);
+            }
+        }
+
+        auto & min_tls_version = _creds->minimum_tls_version();
+        auto & max_tls_version = _creds->maximum_tls_version();
+
+        if (min_tls_version.has_value()) {
+            if (!SSL_CTX_set_min_proto_version(ssl_ctx.get(),
+                tls_version_to_ossl(*min_tls_version))) {
+                throw make_ossl_error(
+                    fmt::format("Failed to set minimum TLS version to {}",
+                        *min_tls_version));
+            }
+        }
+
+        if (max_tls_version.has_value()) {
+            if (!SSL_CTX_set_max_proto_version(ssl_ctx.get(),
+                tls_version_to_ossl(*max_tls_version))) {
+                    throw make_ossl_error(
+                        fmt::format("Failed to set maximum TLS version to {}",
+                            *max_tls_version));
+            }
+        }
+
+        auto get_min_level = [&ssl_ctx]() {
```

Comment:
Nit: Should be "max level" really, as we should consider us setting the max level we can use given the tls version bounds.